### PR TITLE
CP-50194 Implement xenserver_vm provider resource

### PR DESCRIPTION
**Pull request by acefei**
```
source .env && TF_ACC=1 go test ./xenserver/ -v   -timeout 120m
=== RUN   TestAccNetworkDataSource
--- PASS: TestAccNetworkDataSource (0.86s)
=== RUN   TestAccVlanResource
--- PASS: TestAccVlanResource (2.78s)
=== RUN   TestAccNICDataSource
--- PASS: TestAccNICDataSource (0.84s)
=== RUN   TestAccPifDataSource
--- PASS: TestAccPifDataSource (0.77s)
=== RUN   TestAccSnapshotResource
--- PASS: TestAccSnapshotResource (6.37s)
=== RUN   TestAccSRDataSource
--- PASS: TestAccSRDataSource (0.76s)
=== RUN   TestAccNFSResource
--- PASS: TestAccNFSResource (8.84s)
=== RUN   TestAccSRResourceLocal
--- PASS: TestAccSRResourceLocal (3.58s)
=== RUN   TestAccSRResourceShared
--- PASS: TestAccSRResourceShared (8.27s)
=== RUN   TestAccVDIResource
--- PASS: TestAccVDIResource (8.76s)
=== RUN   TestAccVMDataSource
--- PASS: TestAccVMDataSource (2.42s)
=== RUN   TestAccVMResource
--- PASS: TestAccVMResource (4.73s)
PASS
ok      terraform-provider-xenserver/xenserver  48.988s
```

### CODE REVIEW

File: xenserver/vm_utils.go [hunk 1]
```diff
@@ -439,9 +474,11 @@ func setOtherConfigFromPlan(ctx context.Context, session *xenapi.Session, plan v
 
 	originalTFOtherConfigKeys := vmOtherConfig["tf_other_config_keys"]
 	// To compare originalTFOtherConfigKeys with tfOtherConfigKeys, if the key is not in tfOtherConfigKeys, delete it
-	for key := range vmOtherConfig {
-		if !strings.Contains(tfOtherConfigKeys, key+",") && strings.Contains(originalTFOtherConfigKeys, key+",") {
+	originalKeys := strings.Split(originalTFOtherConfigKeys, ",")
+	for _, key := range originalKeys {
+		if !strings.Contains(tfOtherConfigKeys, key+",") {
 			delete(vmOtherConfig, key)
+			tflog.Debug(ctx, "Deleted key from other config: "+key)
```
Comment (xueqingz) [inline xenserver/vm_utils.go:481]: Deleted -> delete
or just remove this debug

Comment (acefei) [reply xenserver/vm_utils.go:481]: removed

File: xenserver/vbd_utils.go [hunk 2]
```diff
@@ -54,16 +54,16 @@ func VBDSchema() map[string]schema.Attribute {
 
 func setVBDDefaults(vbd *vbdResourceModel) {
 	// Work around for https://github.com/hashicorp/terraform-plugin-framework/issues/726
-	if vbd.Mode.IsUnknown() {
+	if vbd.Mode.IsUnknown() || vbd.Mode.IsNull() {
```
Comment (xueqingz) [inline xenserver/vbd_utils.go:57]: why add " vbd.Mode.IsNull() " as Mode is a computed value , never be null?

Comment (acefei) [reply xenserver/vbd_utils.go:57]: but hard_drive is option attribute, it would be not set in main.tf.

File: xenserver/vbd_utils.go [hunk 3]
```diff
@@ -112,7 +117,7 @@ func createVBD(vbd vbdResourceModel, vmRef xenapi.VMRef, session *xenapi.Session
 	return nil
 }
 
-func createVBDs(ctx context.Context, data vmResourceModel, vmRef xenapi.VMRef, session *xenapi.Session) error {
+func createVBDs(ctx context.Context, session *xenapi.Session, vmRef xenapi.VMRef, data vmResourceModel, vbdType xenapi.VbdType) error {
```
Comment (xueqingz) [inline xenserver/vbd_utils.go:120]: remove " vbdType xenapi.VbdType" form params and change to  like:
`err := createVBD(session, vmRef, vbd, xenapi.VbdTypeDisk)`

File: xenserver/vbd_utils.go [hunk 4]
```diff
@@ -197,3 +202,101 @@ func updateVBDs(ctx context.Context, plan vmResourceModel, state vmResourceModel
 
 	return nil
 }
+
+func getVDIRefFromISOName(session *xenapi.Session, isoName string) (string, error) {
```
Comment (xueqingz) [inline xenserver/vbd_utils.go:206]: change function name here , eg 
 `getVDIUUIDFromISOName`

File: xenserver/vm_utils.go [hunk 5]
```diff
@@ -168,26 +181,48 @@ func VMSchema() map[string]schema.Attribute {
...
+			Optional:            true,
+			Computed:            true,
+			Validators: []validator.String{
+				stringvalidator.RegexMatches(regexp.MustCompile(`^[cdn]{1,3}$`), "the value is combination string of ['c', 'd', 'n']"),
+			},
+		},
+		"cdrom": schema.StringAttribute{
+			MarkdownDescription: "The VDI Name in ISO Library to attach to the virtual machine, if not set, use the default value from the template.",
+			Optional:            true,
+			Computed:            true,
```
Comment (xueqingz) [inline xenserver/vm_utils.go:211]: on design , is value is not computed.

File: xenserver/vm_utils.go [hunk 6]
```diff
@@ -675,13 +763,102 @@ func updateVMCPUs(session *xenapi.Session, vmRef xenapi.VMRef, data vmResourceMo
 	return nil
 }
 
+func updateBootOrder(session *xenapi.Session, vmRef xenapi.VMRef, plan vmResourceModel) error {
+	var err error
+	bootOrder := plan.BootOrder.ValueString()
+	if bootOrder == "" {
```
Comment (xueqingz) [inline xenserver/vm_utils.go:769]: As there's a Validators, is this possible?

File: xenserver/vm_utils.go [hunk 7]
```diff
@@ -675,13 +763,102 @@ func updateVMCPUs(session *xenapi.Session, vmRef xenapi.VMRef, data vmResourceMo
...
+		return err
+	}
+
+	// if bootOrder is not empty, set the value to the VM
+	err = setHVMBootParams(session, vmRef, map[string]string{"order": bootOrder})
+	if err != nil {
+		return errors.New(err.Error())
```
Comment (xueqingz) [inline xenserver/vm_utils.go:776]: return err

File: xenserver/vm_utils.go [hunk 8]
```diff
@@ -675,13 +763,102 @@ func updateVMCPUs(session *xenapi.Session, vmRef xenapi.VMRef, data vmResourceMo
...
+	}
+
+	return err
+}
+
+func updateBootMode(session *xenapi.Session, vmRef xenapi.VMRef, plan vmResourceModel) error {
+	var err error
+	bootMode := plan.BootMode.ValueString()
+	// if bootmode is empty, get the value from the template
+	if bootMode == "" {
```
Comment (xueqingz) [inline xenserver/vm_utils.go:786]: I think it's don't need

File: xenserver/vm_utils.go [hunk 9]
```diff
@@ -675,13 +763,102 @@ func updateVMCPUs(session *xenapi.Session, vmRef xenapi.VMRef, data vmResourceMo
...
+	if bootMode == "uefi_security" {
+		bootMode = "uefi"
+		err = setPlatform(session, vmRef, map[string]string{"secureboot": "true"})
+		if err != nil {
+			return errors.New(err.Error())
+		}
+	} else {
+		err = setPlatform(session, vmRef, map[string]string{"secureboot": "false"})
+		if err != nil {
+			return errors.New(err.Error())
```
Comment (xueqingz) [inline xenserver/vm_utils.go:800]: return err

File: xenserver/vm_utils.go [hunk 10]
```diff
@@ -675,13 +763,102 @@ func updateVMCPUs(session *xenapi.Session, vmRef xenapi.VMRef, data vmResourceMo
...
+	if bootMode == "" {
+		return err
+	}
+
+	// if bootmode is not empty, set the value to the VM
+	if bootMode == "uefi_security" {
+		bootMode = "uefi"
+		err = setPlatform(session, vmRef, map[string]string{"secureboot": "true"})
+		if err != nil {
+			return errors.New(err.Error())
```
Comment (xueqingz) [inline xenserver/vm_utils.go:795]: return err

File: xenserver/vbd_utils.go (hunk 4 above)
Comment (acefei) [reply xenserver/vbd_utils.go:206]: I don't think Ref is like UUID need to be capital.

File: xenserver/vm_utils.go (hunk 6 above)
Comment (acefei) [reply xenserver/vm_utils.go:769]: it is optional attribute, if not provided, ValueString() returns "".

File: xenserver/vm_utils.go [hunk 11]
```diff
@@ -675,13 +763,102 @@ func updateVMCPUs(session *xenapi.Session, vmRef xenapi.VMRef, data vmResourceMo
...
+}
+
+func updateBootMode(session *xenapi.Session, vmRef xenapi.VMRef, plan vmResourceModel) error {
+	var err error
+	bootMode := plan.BootMode.ValueString()
+	// if bootmode is empty, get the value from the template
+	if bootMode == "" {
+		return err
+	}
+
```
Comment (xueqingz) [inline xenserver/vm_utils.go:789]: What about  
```
secureboot := "false"
if bootMode == "uefi_security" {
		bootMode = "uefi"
               secureboot = "true"
}
err = setPlatform(session, vmRef, map[string]string{"secureboot": secureboot})
if err != nil {
	return err
}


```

File: xenserver/vm_utils.go (hunk 7 above)
Comment (acefei) [reply xenserver/vm_utils.go:776]: There is an annoy golangci-lint for errwrap.

File: xenserver/vm_utils.go (hunk 8 above)
Comment (acefei) [reply xenserver/vm_utils.go:786]: see above.

File: xenserver/vm_utils.go (hunk 9 above)
Comment (acefei) [reply xenserver/vm_utils.go:800]: see above.

File: xenserver/vm_utils.go (hunk 10 above)
Comment (acefei) [reply xenserver/vm_utils.go:795]: see above.

File: xenserver/vbd_utils.go [hunk 12]
```diff
@@ -197,3 +202,101 @@ func updateVBDs(ctx context.Context, plan vmResourceModel, state vmResourceModel
...
+		tflog.Debug(ctx, "---> CDROM is not set, use the default value from the VM template")
+		return nil
+	}
+
+	vmRecord, err := xenapi.VM.GetRecord(session, vmRef)
+	if err != nil {
+		return errors.New(err.Error())
+	}
+
+	stateISOName, ISOVBDRef, err := getISOFromVMRecord(ctx, session, vmRecord)
```
Comment (xueqingz) [inline xenserver/vbd_utils.go:233]: current resource ISO name

File: xenserver/vbd_utils.go [hunk 13]
```diff
@@ -197,3 +202,101 @@ func updateVBDs(ctx context.Context, plan vmResourceModel, state vmResourceModel
...
+
+func getVDIRefFromISOName(session *xenapi.Session, isoName string) (string, error) {
+	var vdiUUID string
+	vdiRecords, err := xenapi.VDI.GetAllRecords(session)
+	if err != nil {
+		return vdiUUID, errors.New(err.Error())
+	}
+
+	for _, vdiRecord := range vdiRecords {
+		if vdiRecord.NameLabel == isoName {
```
Comment (xueqingz) [inline xenserver/vbd_utils.go:214]: multiple iso name ?

File: xenserver/vbd_utils.go (hunk 3 above)
Comment (acefei) [reply xenserver/vbd_utils.go:120]: Keep the code for the future expansion

File: xenserver/vm_resource_test.go [hunk 14]
```diff
@@ -48,7 +50,7 @@ resource "xenserver_vm" "test_vm" {
   	"flag" = "1"
   }
 }
-`, name_label, memory, vcpu, bootable, mode, mac, device)
+`, name_label, memory, vcpu, boot_mode, boot_order, bootable, mode, mtu, mac, device)
```
Comment (xueqingz) [inline xenserver/vm_resource_test.go:53]: remove  mtu

File: xenserver/vm_utils.go (hunk 7 above)
Comment (xueqingz) [reply xenserver/vm_utils.go:775]: no, we don't need error.New here.

**Review by xueqingz — APPROVED** [review]

**Review by xihuan-citrix — APPROVED** [review]

**Pull request merged**